lib/settings-formats.nix

```diff
@@ -0,0 +1,132 @@
+{ lib }:
+rec {
+
+  # unset :: Value
+  # This value represents nothingness, as if `mkIf false` was used in a
+  # definition.
+  unset = rec {
+    _nixos_unset_value = throw
+      ( "You tried to access an option value that is unset. "
+      + "Use `lib.formats.getOrDefault` on the value before using it.");
+
+    __toString = _nixos_unset_value;
+  };
+
+  # getOrDefault :: EmptyFreeValue -> Value -> EmptyFreeValue
+  # A function for deeply stripping all unset values from the second argument
+  # Falls back to the first argument if the second argument itself is unset
+  getOrDefault =
+    let
+      unsetAttrNames = lib.attrNames unset;
+      isUnset = value: lib.isAttrs value && lib.attrNames value == unsetAttrNames;
+    in fallback: value:
+    if isUnset value then
+      fallback
+    else if lib.isAttrs value then
+      # Since we filtered unset values already we know the fallback won't be hit
+      lib.mapAttrs (name: getOrDefault null)
+        (lib.filterAttrs (name: value: ! isUnset value)
+        # In case the unset attribute set is merged with other attribute sets
+        (builtins.removeAttrs value unsetAttrNames))
+    else if lib.isList value then
+      map (getOrDefault null) (lib.filter (value: ! isUnset value) value)
+    else value;
+
+
+  /*
+
+  Every following entry represents a format for program configuration files
+  used for `settings`-style options (see https://github.com/NixOS/rfcs/pull/42).
+  Each entry should look as follows:
+
+    <format> = {
+
+      # The module system type most suitable for representing such a format
+      # In order to allow attribute sets to depend on themselves, use
+      # `lazyAttrsOf (<type> // { emptyValue.value = unset; })` instead of
+      # `attrsOf <type>`
+      # The description should be overwritten for recursive types
+      type = ...;
+
+      # generate :: Pkgs -> Name -> Value -> Path
+      # A function for generating a file with a value of such a type
+      # Empty values need to be stripped first with the `getOrDefault` function
+      # and an appropriate fallback
+      generate = ...;
+
+    });
+  */
+
+
+  json = {
+
+    type = with lib.types; let
+      valueType = nullOr (oneOf [
+        bool
+        int
+        float
+        str
+        (lazyAttrsOf valueType)
+        (listOf valueType)
+      ]) // {
+        description = "JSON value";
+        emptyValue.value = unset;
+      };
+    in valueType;
+
+    generate = pkgs: name: value: pkgs.runCommandNoCC name {
+      nativeBuildInputs = [ pkgs.jq ];
```

Comment:
if you use `pkgs`, this cannot live in `lib/` anymore; anything involving build-time should like in `pkgs/build-support`.

Though maybe it could be split, so that the pure nix parts still live in `lib/`.

Reply:
Edit: saw the `generate` structure. If we need pkgs for these functions, I’m really not sure this is a good fit for `lib/`.

Reply:
Also thought about this, but I'm not entirely sure what to do, because the `type` part doesn't need `pkgs` and it makes for sense for it to be in `lib`, unlike the `generate` part. However those two parts should be very close together, so splitting them in `pkgs` and `lib` isn't optimal either. I guess putting both in `pkgs.formats.{json,ini,yaml,toml}.{type,generate}` might be the best solution, even though I don't really like it.

Reply:
@Profpatsch I'd rather vote to move `pkgs/build-support/trivial-builders.nix` to `lib`. Let's leave "pure" lib to `<nix/lib>`, not `<nixpkgs/lib>`.

Reply:
@danbst I think a problem with this is that there's no reasonable default `pkgs` to use for library functions that depend on it. The only choice would be to use `import ../. {}`, however then for NixOS, that nixpkgs is imported and evaluated twice, once for `lib`, and once for `pkgs` itself (with potential overlays and stuff).

Reply:
The changes in this PR now live under `nixpkgs/pkgs/pkgs-lib`, which is intended for things that would fit into `nixpkgs/lib` but have a dependency on `pkgs`